### PR TITLE
Fix MapboxTripSession isOffRoute set logic causing missing re-routes

### PR DESCRIPTION
**Pull request by Guardiola31337**
## Description

Fixes `MapboxTripSession` `isOffRoute` `set` logic causing missing re-routes if multiple `OFF_ROUTE` events were sent

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Fix 👇 

> After setting a route, the initial `getStatus` is `invalid`, doesn't reroute, and even after returning to the route line, there are no progress updates (we're constantly `invalid`). 

### Implementation

NN is the source of truth for re-routes so we should trust it and be able to safely remove https://github.com/mapbox/mapbox-navigation-android/blob/432e283abf61e12ca9765143355b7d6b5a609cd4/libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt#L115-L117 that was causing missing re-routes if NN sends 2 consecutive `OFF_ROUTE` events - the SDK is only emitting 1 re-route because the `isOffRoute` state never gets reset.

Also https://github.com/mapbox/mapbox-navigation-android/blob/432e283abf61e12ca9765143355b7d6b5a609cd4/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L606-L608 was interrupting on-going re-route requests as these could take some time and the `status.routeState` could change before the route being fetched. Again, if we can guarantee that NN sends `OFF_ROUTE` events correctly (only once when NN detects so) we can safely remove this `else` branch and all re-routes should be handled once and 👌 

## Testing

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make 1.0-core-update-api` (Core) / `$> make 1.0-ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

cc @SiarheiFedartsou @mskurydin

### CODE REVIEW

**Comment by RingerJK** [comment]
I think it's not a best solution, because we will receive spam of off/on-route events.
IMO we're missing events if we were off-route and set route, but we're not at this route actually, so my suggestion: 
after `setRoute` force to set to observers actual value of `offRoute`. Smth like
```
override var route: DirectionsRoute? = null
        set(value) {
            field = value
            ioJobController.scope.launch {
                val navStatus = navigator.setRoute(value)
                mainJobController.scope.launch {
                   offRouteObservers.forEach{ it.onOffRouteStateChanged(navStatus.routeState == RouteState.OFF_ROUTE) } 
                }
            }
        }
```
it prevents us of spam and we still might receive events like offRoute = true twice, when route has been set

**Comment by Guardiola31337** [comment]
> because we will receive spam of off/on-route events.

I thought about this as well and you're right. Also thought ways of minimizing the spam if we want to do so:

- Similar to what you suggested - I think with this approach intermediate `OFF_ROUTE` events could be missed as `RouteProgress` updates keep pushing

```kotlin
override var route: DirectionsRoute? = null
        set(value) {
            field = value
            ioJobController.scope.launch {
                navigator.setRoute(value)
                isOffRoute = false
            }
        }
```

- Not notifying if `isOffRoute` was `false` and new value too

```kotlin
private var isOffRoute: Boolean = false
        set(value) {
            if (field == value && !field) {
                return
            }
            field = value
            offRouteObservers.forEach { it.onOffRouteStateChanged(value) }
        }
```

But thinking deeper, isn't pushing the state all the time the purpose of having a reactive observer? 🤔 

Also thought about `OffRouteObserver` https://github.com/mapbox/mapbox-navigation-android/blob/432e283abf61e12ca9765143355b7d6b5a609cd4/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/OffRouteObserver.kt#L5-L14

If we'd like to notify only when we detect an `OFF_ROUTE` event the `offRoute` flag is useless it should be something like

```kotlin
fun onOffRoute()
```

Also thought about pushing the `RouteProgressState` like

```kotlin
fun onRouteProgressStateChanged(state: RouteProgressState)  
```

but that's already available from the `RouteProgress` which made me think: do we need this observer at all? Is it providing any benefits? 🤔 

One thing that I also realized when going through this is that if we want to keep it we should only push `onOffRouteStateChanged` updates when in _Active Guidance_ - currently we're pushing also when in _Free Drive_ as `RouteProgress` updates are emitted all the time as part of a trip session. Going to push a change to address this.

All these thoughts directly relate to https://github.com/mapbox/mapbox-navigation-android/issues/3292

What do you think @RingerJK @kmadsen @LukasPaczos?

**Comment by RingerJK** [comment]
I don't see anything bad, that we're notify about route state, because it's a state of a NEW route, so if developer or sdk(internally) receive the same offRoute event(true or false it's not important here), it a state of new route, that is actually expected. Also we could update `OffRouteObserver` API to determine if it's a new route :

suggestion 1
```
interface OffRouteObserver { 
  fun onOffRouteStateChanged(offRoute: Boolean, uuid: String) // uuid of route
}
```

suggestion 2 
```
interface OffRouteObserver { 
  fun onOffRouteStateChanged(offRoute: Boolean)
  fun onNewRouteHasBeenSet(offRoute: Boolean) // called only once when new route is set
}
```

**Comment by Guardiola31337** [comment]
@RingerJK 
> I don't see anything bad, that we notify about route state, because it's a state of a NEW route, so if developer or sdk(internally) receive the same offRoute event(true or false it's not important here), this's a state of new route, that is actually expected. Also we could update `OffRouteObserver` API to determine if it's a new route :
>
>suggestion 1
>```
>interface OffRouteObserver { 
>  fun onOffRouteStateChanged(offRoute: Boolean, uuid: String) // uuid of route
>}
>```
>
>suggestion 2 
>```
>interface OffRouteObserver { 
>  fun onOffRouteStateChanged(offRoute: Boolean)
>  fun onNewRouteHasBeenSet(offRoute: Boolean) // called only once when new route is set
>}
>```

These are definitely suggestions trying to address the complaints from https://github.com/mapbox/mapbox-navigation-android/issues/3292 Also it's worth noting that these are two separate things that currently aren't connected - `OffRouteObserver` notifies when `RouteProgressState` "changes" but unrelated to the route about to be retrieved `RerouteController` i.e. `OFF_ROUTE` event is received and then a `reroute` request is emitted and afterwards the new route is fetched, in that particular order. We can keep the discussion on how to extend it in that other ticket.

On the other topic, if the idea is to only notify when `isOffRoute` changes we might need to add logic around the `set`ter and emit the updates accordingly i.e. what happens if the `isOffRoute` is `true` and the new value is `true`? Same question for all combinations possible - there aren't that many, only 4 😅 

That'd be expanding 👇 idea and cover all cases.

> - Not notifying if `isOffRoute` was `false` and new value too
>
>```kotlin
>private var isOffRoute: Boolean = false
>        set(value) {
>            if (field == value && !field) {
>                return
>            }
>            field = value
>            offRouteObservers.forEach { it.onOffRouteStateChanged(value) }
>        }
>```

It's another option.

**Comment by RingerJK** [comment]
> On the other topic, if the idea is to only notify when isOffRoute changes we might need to add logic around the setter and emit the updates accordingly i.e. what happens if the isOffRoute is true and the new value is true? Same question for all combinations possible - there aren't that many, only 4 😅

As I mentioned above it's ok if `offRouteObserver` is receiving two times `true` or two times `false` because these states are for different routes, so it's expected behaviour 👍 

> ```kotlin
>    private var isOffRoute: Boolean = false
>       set(value) {
>            if (field == value && !field) {
>                return
>            }
>            field = value
>            offRouteObservers.forEach { it.onOffRouteStateChanged(value) }
>        }
> ```

in this case we don't receive true state twice, in case of a **new** route. Also it will spam `offRoute=true`

**Comment by Guardiola31337** [comment]
@RingerJK 
> in this case we don't receive true state twice, in case of a **new** route.

What do you mean by `in case of a new route`? This is not connected to the route state so I believe that's expected.

> Also it will spam `offRoute=true`

Yeah, as I was saying if we trust NN it should be fine to "spam" (it'll only be received when NN reports `OFF_ROUTE` which is the source of truth) and SDK will notify back. Again this is irrespective of the state of the route.

If we want to consider / add the route state into play we need to verify that 👇 is not an issue.

> - Similar to what you suggested - I think with this approach intermediate `OFF_ROUTE` events could be missed as `RouteProgress` updates keep pushing
>
>```kotlin
>override var route: DirectionsRoute? = null
>        set(value) {
>            field = value
>            ioJobController.scope.launch {
>                navigator.setRoute(value)
>                isOffRoute = false
>            }
>        }
>```

cc @LukasPaczos

**Comment by RingerJK** [comment]
That will make infinitive loop of reroutes ⬇️ 
>```kotlin
>override var route: DirectionsRoute? = null
>        set(value) {
>            field = value
>            ioJobController.scope.launch {
>                navigator.setRoute(value)
>                isOffRoute = false
>            }
>        }
>```

I think we should be fine to set _real_ state just after `setRoute`. To avoid misunderstanding what route is on/offroute we might extend `OffrouteObserver` API 👍

**Comment by RingerJK** [comment]
That's not fear, because we might be offroute all the time, so it's false positive
>```kotlin
>override var route: DirectionsRoute? = null
>        set(value) {
>            field = value
>            ioJobController.scope.launch {
>                navigator.setRoute(value)
>                isOffRoute = false
>            }
>        }
>```

I think we should be fine to set _real_ state just after `setRoute`. To avoid misunderstanding what route is on/offroute we might extend `OffrouteObserver` API 👍

**Comment by Guardiola31337** [comment]
This is ready for another round of :eyes: @RingerJK @LukasPaczos

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt [hunk 1]
```diff
@@ -79,11 +82,19 @@ class MapboxTripSession(
...
         set(value) {
             field = value
             ioJobController.scope.launch {
-                navigator.setRoute(value)
+                mutex.withLock {
+                    val status = navigator.setRoute(value)
+                    // Could NN send OFF_ROUTE as a result of setRoute? If not (NN only sends INVALID or INITIALIZED), isOffRoute = false should be safe
+                    // Would that be the case with spamming OFF_ROUTE approach? If so, isOffRoute will be always true which may cause issues
+                    // From a NN point, is it safe to assume that setting a new route means never OFF_ROUTE?
+                    isOffRoute = status.routeState == RouteState.OFF_ROUTE
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt:96]:
```suggestion
   mainJobController.scope.launch {
                    isOffRoute = status.routeState == RouteState.OFF_ROUTE
    }
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 2]
```diff
@@ -604,9 +604,10 @@ class MapboxNavigation(
     private fun createInternalOffRouteObserver() = object : OffRouteObserver {
         override fun onOffRouteStateChanged(offRoute: Boolean) {
             if (offRoute) {
+                if (rerouteController?.state == RerouteState.FetchingRoute) {
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:607]: I imagine this logic should be moved into RerouteController

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 3]
```diff
@@ -68,11 +71,16 @@ internal class MapboxTripSession(
...
+    private val mutex = Mutex()
+
     override var route: DirectionsRoute? = null
         set(value) {
             field = value
             ioJobController.scope.launch {
-                navigator.setRoute(value)
+                mutex.withLock {
+                    navigator.setRoute(value)
+                    isOffRoute = false
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:82]: `isOffRoute` must be changed via Main thread only, or need sync mechanism

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 2 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:607]: Agreed :+1:

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt [hunk 4]
```diff
@@ -115,25 +133,7 @@ internal class MapboxRerouteController(
...
-        mainJobController.scope.launch {
-            rerouteStateObserver.onRerouteStateChanged(state)
-        }
-        return observers.add(rerouteStateObserver)
-    }
-
-    override fun unregisterRerouteStateObserver(rerouteStateObserver: RerouteController.RerouteStateObserver): Boolean {
-        return observers.remove(rerouteStateObserver)
+    private companion object {
+        const val TAG = "MapboxRerouteController"
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: why did you move companion to bottom of the file 🤔

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: Did some method re-ordering and moved it to bottom, we should consolidate that across the codebase - currently in some classes are at the top and in some others at the bottom ¯\\\_(ツ)\_/¯

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 5]
```diff
@@ -604,9 +604,8 @@ class MapboxNavigation(
     private fun createInternalOffRouteObserver() = object : OffRouteObserver {
         override fun onOffRouteStateChanged(offRoute: Boolean) {
             if (offRoute) {
-                reroute()
-            } else {
                 rerouteController?.interrupt()
+                reroute()
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:608]: seems that in case when we/customer call `reroute` we alway must interrupt current reroute call, so I'd suggest to move interrupt under `reroute` in `MapboxRerouteController`.
So `MapboxRerouteController.kt` will look like
```
private fun request(routeOptions: RouteOptions) {
  interrupt()
  ...
}
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt (hunk 4 above)
Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: imo companions better place is top of the class, just after all properties. 
btw Google samples [are driving with this approach ](https://github.com/android/user-interface-samples/tree/master/EmojiCompatKotlin/app/src/main/java/com/example/android/emojicompat)

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: On the contrary, Kotlin suggests to put them at the end ¯\\\_(ツ)\_/¯ https://kotlinlang.org/docs/reference/coding-conventions.html#class-layout

In any case, I don't personally have a strong opinion only looking for consistency across the codebase.

Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:137]: yeah, kotlin guide likes to do it :)

**Comment by Guardiola31337** [comment]
Addressed PR comments, this is ready for another round of 👀 - after getting your blessing will rebase the commit history and we should be good to merge here 🚀 @LukasPaczos @RingerJK

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 6]
```diff
@@ -414,7 +423,9 @@ internal class MapboxTripSession(
             launch {
                 updateEnhancedLocation(status.enhancedLocation, status.keyPoints)
                 updateRouteProgress(status.routeProgress)
-                isOffRoute = status.offRoute
+                mutex.withLock {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:426]: Wouldn't moving this lock a level or 2 up also resolve https://github.com/mapbox/mapbox-navigation-android/issues/3383?

It can be addressed in a follow-up, just something to think about.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/reroute/MapboxRerouteControllerTest.kt [hunk 7]
```diff
@@ -221,6 +221,41 @@ class MapboxRerouteControllerTest {
         }
     }
 
+    @Test
```
Comment (RingerJK) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/reroute/MapboxRerouteControllerTest.kt:224]: ❤️

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt [hunk 8]
```diff
@@ -39,12 +39,9 @@ internal class MapboxRerouteController(
             observers.forEach { it.onRerouteStateChanged(field) }
         }
 
-    private companion object {
-        const val TAG = "MapboxRerouteController"
-    }
-
     // current implementation ignore `onNewRoutes` callback because `DirectionsSession` update routes internally
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:42]:
```suggestion
    // current implementation ignore `routesCallback ` callback because `DirectionsSession` update routes internally
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt (hunk 6 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:426]: Yeah when I saw recent https://github.com/mapbox/mapbox-navigation-android/issues/3383 was wondering the same. Will follow up in a separate PR 👍 Thanks for the heads up!

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:426]: 👍  haven't looked into kotlin's Mutex much until now, but it is pretty cool because it is a suspending function. Meaning, you can call it from the main thread and it won't block it. Coroutines can be so neat!

https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.sync/-mutex/lock.html
"In a blocking world you'd typically use synchronized or ReentrantLock for that. Coroutine's alternative is called Mutex. It has lock and unlock functions to delimit a critical section. The key difference is that Mutex.lock() is a suspending function. It does not block a thread."

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:426]: But +1 to what lukas is saying, moving it up a level will likely resolve #3383. Curious how long it takes to complete though, critical sections should be lock minimal computation
